### DISCUSSION TITLE
How to return a file of loss and accuracy for training ?

### DISCUSSION THREAD

**Discussion by holiday01** [NVIDIA/NVFlare#771] [Q&A]
Is the nvflare api can return the loss and accuracy for training by the admin host? I want to plot the learning curve.

**Comment by YuanTingHsieh** [comment]
Hi @holiday01 ,

Are you asking, after your job is finished, how can you get the training model/loss/accuracy?

When a user do FL, each client site will have its own model.
You can save that using your own custom executor, and you can evaluate using that model to find out loss and accuracy of those local model.

A server side can also save a global model.

Please refer to this documentation on end-to-end: https://nvflare.readthedocs.io/en/2.1.1/examples/hello_pt.html

**Reply by holiday01** [reply]
Are they returned to the clients? How do the loss/accuracy values return to the server?

**Reply by holgerroth** [reply]
You might be interested in NVFlare's tensorboard streaming capabilities. See an example [here](https://github.com/NVIDIA/NVFlare/tree/dev/examples/cifar10#5-streaming-tensorboard-metrics-to-the-server).

**Reply by holiday01** [reply]
Thank @holgerroth and I knew it. I want to get to know more about NVFlare API development, such that we can use "FLAdminAPIRunner" to send a job. What is API or code I can return acc/loss of the cross-validation to the server or admin?

**Reply by holgerroth** [reply]
If you don't want to stream the loss/acc values during training as in the example above, you can also just modify your Executor to add a loss/acc in the shareable that's returned to the server after executing a task.

For submitting jobs with `FLAdminAPIRunner`, have a look [here](https://github.com/NVIDIA/NVFlare/blob/dev/examples/cifar10/submit_job.py).

**Reply by holiday01** [reply]
Thank @holgerroth. I knew that we can get the loss/acc by "log.info" from the sites. Could you give me an example add "log.info" in the shareable.

**Reply by holgerroth** [reply]
One way is to  add the loss/acc information to the DXO that's being returned by the training logic. For example, the cifar10 Learner could adds meta properties [here](https://github.com/NVIDIA/NVFlare/blob/d2ab5b524c8310a3fd4a4b7b0fe88a9708c74427/examples/cifar10/pt/learners/cifar10_learner.py#L334).

**Reply by holgerroth** [reply]
Your workflow running on the server then has access to that meta prop when [receiving](https://github.com/NVIDIA/NVFlare/blob/d2ab5b524c8310a3fd4a4b7b0fe88a9708c74427/nvflare/app_common/workflows/scatter_and_gather.py#L195) the result from the trainer. You could also use a filter to process that result on the server if preferable.
